### ISSUE TITLE
On update only triggered within 10 seconds of client startup

### ISSUE THREAD

**Issue by sachssem** [Bannerets/tdl#158]
Hey, the clients on update method is only called within the first 10 seconds after starting it. Afterwards it just stops emitting any updates. 

At first this was only happening on my server, but while I was trying to debug it over the last week or so it started happening on my local machine too. At least I could further analyse the problem this way and to me it seems like tdlib is receiving the updates properly when inspecting it's logs with verbosity level 4. 
The UpdatesManager of tdlib receives `updateNewChannelMessage` and `updateNewMessage` updates, but as mentioned before `client.on('update', (update) => {})` isn't called.

I set up a second tdl project with another account -> same problem. Now I have tried a basic setup with the first account with pytdbot and it works perfectly fine...

Versions (also tried a self compiled libtdjson):
    "prebuilt-tdlib": "^0.1008030.0",
    "tdl": "^8.0.0"

**Comment by Bannerets** [comment]
Interesting. For what it's worth, a similar issue with missing updates can happen if you create multiple sessions with the same bot account (although if you see the updates in the tdlib log with verbosityLevel 4, that shouldn't be the culprit here).

May I have some steps or code to reproduce that? Do you have the listener on `error`? Do you have multiple `update` listeners? If a listener fails with error, then subsequent listeners won't be called. Once you stop receiving the updates, can you still receive responses to `invoke` requests fine?

**Comment by sachssem** [comment]
I'm using user accounts, not bot accounts. In another Issue of tdlib I've read that up to 10 connections shouldn't be a problem, I am only running one instance of tdl + mobile app (+ web client sometimes).

I can't say how to reproduce this issue, in my setup it's happening with the most basic setup possible:
```
tdl.configure({ tdjson: './libtdjson.dylib', verbosityLevel: 4 })
export const client = tdl.createClient({
  apiId: API_ID,
  apiHash: API_HASH,
})

client.on('error', console.error)
client.on('update', (update) => {
  if (update._ === 'updateNewMessage') console.log('------>updateNewMessage<------')
  else console.log(update._)
})

async function main() {
  await client
    .login({
      async getPhoneNumber(retry) {
        if (retry) throw new Error('Invalid phone number')
        return PHONE_NUMBER
      },
    })
    .then(() => {
      console.log('logged in')
    })
}

main()
```

When I start it the usual amount of updates is received and I can see corresponding updates when coming online and sending a message with my second account:

updateUser
updateUserStatus
updateUserStatus
updateUserStatus
updateUser
updateChatReadInbox
------>updateNewMessage<------
updateChatLastMessage
updateUserStatus


10-20 seconds later (seems like it stops working when I do not write a message within 10-20 secs of the last one) nothing is printed to stdout anymore, not one single update of any kind.

I've never seen a tdlib error in the logs (guess it would be red), only information and debug messages (blue and white).

With
```
async function invokeSmth() {
  console.log('invokeSmth')
  await client.invoke({ _: 'getMe' }).then((res) => console.log(res))
}
```

and
```
setTimeout(() => invokeSmth(), 1000)
setTimeout(() => invokeSmth(), 30000)
```

it works as expected the first time and prints the user object via the on message handler, but after 30 secs only 'invokeSmth' gets printed.

tdlib logs of the second invokeSmth call @ verbosity level 4:
```
invokeSmth
[ 3][t 4][1718560790.986262083][Td.cpp:2811][#1][!Td][&td_requests]     Receive request 4: getMe {
}
[ 3][t 4][1718560790.986383914][Td.cpp:3961][#1][!Td][&td_requests]     Sending result for request 4: user {
  id = 501...
  [The usual user object...]
}
[ 4][t 4][1718560790.986446142][Td.cpp:3183][#1][!Td]   Decrease request actor count to 1


[ 4][t 4][1718560795.546506166][AuthData.cpp:122][#1][!Session:2:main]  Create identifier for message 0x666f281b0347705c at 1493391.899364
[ 4][t 4][1718560795.546532154][AuthData.cpp:122][#1][!Session:2:main]  Create identifier for message 0x666f281b0358e504 at 1493391.899392
[ 4][t 4][1718560795.546535015][AuthData.cpp:122][#1][!Session:2:main]  Create identifier for message 0x666f281b0359020c at 1493391.899395
[ 4][t 4][1718560795.546658039][BufferedFd.h:215][#1][!Session:2:main]  Flush write: +220B[left:0B]
[ 4][t 4][1718560795.546667098][SessionConnection.cpp:1140][#1][!Session:2:main]        Last pong was in 29.945872/29.945872, last read was in 29.945913/29.945913, RTT = 2.000000, ping timeout = 137.098417, read timeout = 137.098417, flush packet in 62.098253, wakeup in 62.098253
[...]
```
don't contain `Sending update: updateUser` between `Receive request` and `Sending result` steps, but still log the user object via tdlib logging, just tdls`client.on('update', (update) => {}` isn't triggered anymore...

**Comment by sachssem** [comment]
I should maybe add that it was working fine since February and the issues started with the tdlib 8.0.0 update. 
1-2 days after the update my client stopped working and I had to update tdlib and tdl to make it work again for some days. Then the problems started showing up like I mentioned earlier, first on my server then on my local machine too.

**Comment by Bannerets** [comment]
Thanks for the information. That looks weird, especially given that you receive responses to invoke. Unfortunately, I can still get updates even long after 10 seconds and after a long pause in updates (although I use a bot account) and can't reproduce that. I may suggest to launch the app with the `DEBUG=tdl*` env variable (which will print all tdlib objects received by tdl besides the other things), along with verbosityLevel 4 or 3 (posting this log here with received updates may be helpful). Also, what's your Node.js version?

> seems like it stops working when I do not write a message within 10-20 secs of the last one

By the way, afaik TDLib may not send all chat-related updates if you don't call `openChat` for that chat beforehand.

> The UpdatesManager of tdlib receives updateNewChannelMessage and updateNewMessage updates

I think the log you want to look for is the td_requests one (`[&td_requests]	Sending update`), which prints when something is sent to the app. UpdatesManager is internal to TDLib.

**Comment by Bannerets** [comment]
Reading your [issue](https://github.com/tdlib/td/issues/2930) in the TDLib repository:
> - Then somehow most groups and channels stop receiving updates, only a few messages get delivered in the course of a day (although it should be multiple hundreds)
> - The weird thing is that all of my own messages which I send to myself (or via another account) are received instantly. No matter if they are sent as direct message, in a group or a channel

This makes me think that `openChat` may indeed be the culprit and that TDLib just doesn't send the updates to tdl (you should check blue `td_requests` logs for that).

> In another issue there was a similar problem when someone received updates with a huge delay, compared with TG web client or mobile app - and the solution was to update tdlib.

Updates, especially in large chats, can be delayed without `openChat` (see e.g. https://github.com/tdlib/td/issues/2853)

**Comment by sachssem** [comment]
I have tried calling `openChat` and `closeChat` on all chats periodically, one after another, but the channels still don't receive updates.  My own group works fine, even after hours of the instance running any new message in it is  printed instantly.

When testing it with `DEBUG=tdl*` and verbosity set to 3 I found out something new: 

A new message directly after startup gives the following shortened output:
- `[!UpdatesManager]	Receive updates`
- `[!UpdatesManager]	Receive from telegram_api::updates pending updateNewChannelMessage`
- `[!UpdatesManager]	Receive from updateNewChannelMessage pending updateNewChannelMessage`
- `[&td_requests]	Sending update: updateNewMessage`
- `[&td_requests]	Sending update: updateChatLastMessage`
- `tdl:client:receive { _: 'updateChatReadInbox', [...] }`
- `tdl:client:receive {_:  'updateNewMessage', [...] }`

Thirty seconds later a second message produces the same output except both `tdl:client:receive ` messages are missing!

Seems to me like TDLib receives all updates properly, they are just not passed to tdl somehow.

**Comment by Bannerets** [comment]
> Seems to me like TDLib receives all updates properly, they are just not passed to tdl somehow.

The question is: does TDLib send those updates to tdl? That is, do you see `[&td_requests] Sending update` with the missing updates in the log? It seems unlikely that tdl emits only a subset of updates (the code definitely doesn't differentiate from which chat an update comes), so the issue should be on the TDLib's side.

Another thing that hypothetically may have an effect is the tdlibParameters that are passed to TDLib. tdl's defaults are [here](https://github.com/Bannerets/tdl/tree/7b4d0327f600b007a3045ace5b492fb97e4c9567#tdlcreateclientoptions-clientoptions--client). pytdbot's defaults are a bit different, for example.

One more thing: I feel that isn't likely to solve the issue, but you may try to use `tdl.configure({ useOldTdjsonInterface: true })` (before creating clients), which will use a different interface to communicate with TDLib.

**Comment by sachssem** [comment]
Yes it does send those.

`[&td_requests] Sending update: updateNewMessage` is in the logs of both messages, but only the first one which was delivered has `tdl:client:receive` after that, followed by my updateNewMessage log. 

I haven’t set any special parameters for TDLib:
```
tdl.configure({ tdjson: getTdjson(), verbosityLevel: 4, useOldTdjsonInterface: false })
export const client = tdl.createClient({
  apiId: API_ID,
  apiHash: API_HASH,
})
````

Starting with `useOldTdjsonInterface: true` the process exits right away with exit code 0 and only this output:
```
[ 3][t 0][1718645306.315182209][Client.cpp:600] Create client 1
[ 3][t 0][1718645306.315196037][Client.cpp:383] Initialize client 1
[ 3][t 0][1718645306.315349102][Client.cpp:280][&td_requests]   Begin to wait for updates with timeout 10.000000
[ 3][t 4][1718645306.315532207][Td.cpp:2447][#1][!MultiTd]      Create Td with layer 181, database version 14 and version 53 on 4 threads
```

And I'm using Bun 1.1.13 locally and a node-20 image on my server, so I guess it shouldn't be related to the runtime as well.

**Comment by Bannerets** [comment]
> Bun 1.1.13 locally

That is interesting. I had a suspicion that some runtimes or Node.js versions could be garbage-collecting the TSFN that tdl uses to communicate with TDLib (and hence asked about the Node.js version). Looks like I can reproduce something similar in bun. If no updates/responses are received after a while (after exactly 10s probably, which is the timeout for `receive`), the tdl's node addon stops calling TDLib (if used in bun), and nothing at all is received by tdl. That is,

```js
setInterval(async () => {
  console.log('Got id', (await client.invoke({ _: 'getMe' })).id)
}, 30000)
```

prints no output. This library has been made mostly for Node.js and I haven't really extensively tested it in bun (there's a small note in readme about stability in bun). The cause of the issues in bun could be different from my suspicion though.

The Node.js issue quite possibly may be distinct (even still related to `openChat` perhaps, if you tested that only locally in bun?).

Note that weird things like receiving updates only from specific chats are most likely not happening because of tdl<->TDLib interaction issues; nothing at all is received in this case. The library just passes the received updates or responses further, barely looking into their content.

**Comment by Bannerets** [comment]
https://github.com/Bannerets/tdl/issues/158#issuecomment-2171798192

> it works as expected the first time and prints the user object via the on message handler, but after 30 secs only 'invokeSmth' gets printed.

Also, I was reading hastily and missed this bit in that comment. I thought the responses to invoke are still received and printed; but that prints simply when `invoke` is called, without the responses.

**Comment by Bannerets** [comment]
Briefly looking into why this doesn't work in bun, I think, based on the behavior, the `napi_call_threadsafe_function` implementation there doesn't call the `napi_threadsafe_function_call_js ` function at all if the null pointer is passed as data (and if 10 seconds are passed without updates, `receive` returns null), which (afaik) should be an incorrect implementation.

**Comment by sachssem** [comment]
> even still related to `openChat` perhaps, if you tested that only locally in bun?  

Also tested it with node as I deployed it to my server via a cd pipeline, so it was a clean install and everything for sure.

Messages in our own test group were consistently being delivered there over the past days since the issue arose, really weird. I don't know if you have an explanation for this, the only difference I can spot is that we have a group for testing and the other ones are channels, where the updates aren't delivered.

> By the way, you can open multiple chats.  

Good to know.

> This library has been made mostly for Node.js and I haven't really extensively tested it in bun (there's a small note in readme about stability in bun)

> Well, not all Node-API addons work correctly in bun. I can fix this later with a workaround.

Was always using tdl with Node in production, Bun just for development. Awesome that you could improve it's support there and maybe I should look into deno sometime :D

I will test it asap tomorrow, thank you for looking into it, really appreciate your help!

**Comment by sachssem** [comment]
Okay, so I have changed the openChat implementation now:

- set my client state to online via
```
client
      .invoke({
        _: 'setOption',
        name: 'online',
        value: { _: 'optionValueBoolean', value: false },
      })
```
- open my chats in parallel
- periodically close them and switch to offline for a brief duration before going online and opening them again (every 15 min)

And now it seems to work like before and all messages get delivered! :)
I'm just wondering why it was working without these actions for the past months?..

Furthermore I can confirm that `useOldTdjsonInterface` option works with Bun now. Thank you again.
